synapse/handlers/space_summary.py

```diff
@@ -78,35 +83,81 @@ async def get_space_summary(
-        room_queue = deque((_RoomQueueEntry(room_id),))
+        room_queue = deque((_RoomQueueEntry(room_id, ()),))
 
+        # rooms we have already processed
         processed_rooms = set()  # type: Set[str]
 
+        # events we have already processed. We don't necessarily have their event ids,
+        # so instead we key on (room id, state key)
+        processed_events = set()  # type: Set[Tuple[str, str]]
+
         rooms_result = []  # type: List[JsonDict]
         events_result = []  # type: List[JsonDict]
 
         while room_queue and len(rooms_result) < MAX_ROOMS:
             queue_entry = room_queue.popleft()
             room_id = queue_entry.room_id
+            if room_id in processed_rooms:
+                # already done this room
+                continue
+
             logger.debug("Processing room %s", room_id)
-            processed_rooms.add(room_id)
+
+            is_in_room = await self._store.is_host_joined(room_id, self._server_name)
 
             # The client-specified max_rooms_per_space limit doesn't apply to the
             # room_id specified in the request, so we ignore it if this is the
             # first room we are processing.
             max_children = max_rooms_per_space if processed_rooms else None
 
-            rooms, events = await self._summarize_local_room(
-                requester, room_id, suggested_only, max_children
+            if is_in_room:
+                rooms, events = await self._summarize_local_room(
+                    requester, room_id, suggested_only, max_children
+                )
+            else:
+                rooms, events = await self._summarize_remote_room(
+                    queue_entry,
+                    suggested_only,
+                    max_children,
+                    exclude_rooms=processed_rooms,
+                )
+
+            logger.debug(
+                "Query of %s returned rooms %s, events %s",
+                queue_entry.room_id,
+                [room.get("room_id") for room in rooms],
+                ["%s->%s" % (ev["room_id"], ev["state_key"]) for ev in events],
             )
 
             rooms_result.extend(rooms)
-            events_result.extend(events)
 
-            # add any children that we haven't already processed to the queue
-            for edge_event in events:
-                if edge_event["state_key"] not in processed_rooms:
-                    room_queue.append(_RoomQueueEntry(edge_event["state_key"]))
+            # any rooms returned don't need visiting again
+            processed_rooms.update(cast(str, room.get("room_id")) for room in rooms)
+
+            # the room we queried may or may not have been returned, but don't process
+            # it again, anyway.
+            processed_rooms.add(room_id)
+
+            # XXX: is it ok that we blindly iterate through any events returned by
+            #   a remote server, whether or not they actually link to any rooms in our
+            #   tree?
```

Comment:
I'm not really sure of the implications of this, we're not persisting any of this as state, so joining the room should get the proper data?

Reply:
yes it should.

The concern here is that a malicious or buggy remote server could return a load of events (and rooms, for that matter) which are nothing to do with the requested room. That means that we'll end up spending resources looking for rooms we don't actually care about, and will return that data to the client (which might either ignore it, or (incorrectly) add it to the UI). Still, there shouldn't be any lasting effect here.